module/setup.go

```diff
@@ -30,48 +68,66 @@ func setupHeadersForUpdate(
+	}
+	trustedBoundaryHeight, err := GetBoundaryHeight(getHeader, savedLatestHeight, *trustedCurrentForkSpec)
+	if err != nil {
+		return nil, err
+	}
+	trustedBoundaryEpochs, err := trustedBoundaryHeight.GetBoundaryEpochs(*trustedPreviousForkSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	trustedEpochHeight := trustedBoundaryEpochs.CurrentEpochBlockNumber(savedLatestHeight)
 	latestFinalizedHeight := latestFinalizedHeader.GetHeight().GetRevisionHeight()
-	if latestFinalizedHeight < firstUnsavedEpoch {
-		return withTrustedHeight(append(targetHeaders, latestFinalizedHeader), clientStateLatestHeight), nil
+
+	// If the condition is timestamp. we must submit the header with the timestamp
+	nextForkBoundaryTs, nextForkBoundaryHeightMinus1, err := shouldSubmitBoundaryTimestampHeader(getHeader, savedLatestHeight, latestFinalizedHeader.GetHeight().GetRevisionHeight(), forkSpecs)
+	if err != nil {
+		return nil, err
 	}
 
-	trustedEpochHeight := toEpoch(savedLatestHeight)
+	firstUnsaved := trustedEpochHeight + skip
+	for firstUnsaved <= savedLatestHeight {
+		firstUnsaved += skip
+	}
 
-	// Append insufficient epoch blocks
-	for epochHeight := firstUnsavedEpoch; epochHeight < latestFinalizedHeight; epochHeight += constant.BlocksPerEpoch {
-		verifiableEpoch, err := setupNeighboringEpochHeader(ctx, getHeader, queryVerifiableNeighboringEpochHeader, epochHeight, trustedEpochHeight, latestHeight)
+	submittingHeights := makeSubmittingHeights(latestFinalizedHeight, savedLatestHeight, firstUnsaved, nextForkBoundaryTs, nextForkBoundaryHeightMinus1)
+	logger.Debug("submitting heights", "heights", submittingHeights, "trusted height", savedLatestHeight, "trusted epoch", trustedEpochHeight, "first unsaved", firstUnsaved)
+
+	trustedHeight := clientStateLatestHeight.GetRevisionHeight()
+	for _, submittingHeight := range submittingHeights {
+		verifiableHeader, err := setupIntermediateHeader(ctx, queryVerifiableNeighboringEpochHeader, submittingHeight, latestHeight)
 		if err != nil {
 			return nil, err
 		}
-		if verifiableEpoch == nil {
-			logger.Error("[FastFinalityError]", fmt.Errorf("insufficient vote attestation: epochHeight=%d, trustedEpochHeight=%d", epochHeight, trustedEpochHeight))
+		if verifiableHeader == nil {
+			logger.Error("[FastFinalityError]", fmt.Errorf("insufficient vote attestation: submittingHeight=%d, trusted=%d", submittingHeight, trustedHeight))
 			return withTrustedHeight(targetHeaders, clientStateLatestHeight), nil
 		}
-		targetHeaders = append(targetHeaders, verifiableEpoch)
-		trustedEpochHeight = epochHeight
-		logger.Debug("setup epoch header", "height", epochHeight)
+		targetHeaders = append(targetHeaders, verifiableHeader)
+		trustedHeight = submittingHeight
+		logger.Debug("setup epoch header", "trusted", trustedHeight, "height", submittingHeight)
 	}
 	return withTrustedHeight(append(targetHeaders, latestFinalizedHeader), clientStateLatestHeight), nil
 }
 
-func setupNeighboringEpochHeader(
+func setupIntermediateHeader(
```

Comment:
It would be helpful to have a brief explanation about the new functions.

Reply:
I added the coment by https://github.com/datachainlab/ibc-parlia-relay/pull/57/commits/49dcb36c109c350e995a497afb50127e684202c6